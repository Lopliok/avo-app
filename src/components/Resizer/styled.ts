import React, { ReactNode, useState, useEffect } from "react";
import styled, { css } from "styled-components";

interface DragProps {
    variant: "right" | "bottom" | "corner";
}

export const Container = styled.div`
  position: relative;
  background: white;
  user-select: none;
`;

export const Drag = styled.div`
  position: absolute;
  background: grey;
  ${(props: DragProps) => {
        switch (props.variant) {
            case "right":
                return css`
          right: -10px;
          top: 35px;
          width: 10px;
          height: 30px;
          cursor: e-resize;
        `;
            case "bottom":
                return css`
          left: 20px;
          bottom: -10px;
          width: 30px;
          height: 10px;
          cursor: s-resize;
        `;
            case "corner":
                return css`
          right: -10px;
          bottom: -10px;
          width: 10px;
          height: 10px;
        `;
        }
    }}
`;


