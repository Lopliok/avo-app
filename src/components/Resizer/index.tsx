import React, { ReactNode, useEffect, useState } from "react";
import { Container, Drag } from "./styled";

export interface ResizerProps {
  children: ReactNode;
  defWidth: number;
  defHeight: number;
  dragComponents?: {
    DragCorner: React.ReactNode;
    DragSingleDirection: React.ReactNode;
  };
  styles?: {
    container?: {};
    drag?: {
      right: {};
      bottom: {};
      corner: {};
    };
  };
}

type Direction = "vertical" | "horizontal" | "both";

type Start =
  | {
      positionX: number;
      positionY: number;
      width: number;
      height: number;
      direction: Direction;
    }
  | undefined;

const Resizer: React.FunctionComponent<ResizerProps> = ({
  children,
  defHeight,
  defWidth,
  dragComponents,
  styles
}: ResizerProps) => {
  const [width, setWidth] = useState(defWidth);
  const [height, setHeight] = useState(defHeight);

  const [start, setStart] = useState<Start>();

  const onStartDrag = (
    e: React.MouseEvent<HTMLDivElement>,
    direction: Direction
  ) => {
    setStart({
      positionX: e.clientX,
      positionY: e.clientY,
      width,
      height,
      direction
    });
  };

  const dragging = (e: MouseEvent) => {
    if (start) {
      switch (start.direction) {
        case "vertical":
          setWidth(start.width + e.clientX - start.positionX);
          break;
        case "horizontal":
          setHeight(start.height + e.clientY - start.positionY);
          break;
        case "both":
          setWidth(start.width + e.clientX - start.positionX);
          setHeight(start.height + e.clientY - start.positionY);
          break;
      }
    }
  };

  const dragEnd = () => {
    setStart(undefined);
    window.removeEventListener("mousemove", dragging);
    window.removeEventListener("mouseUp", dragEnd);
  };

  useEffect(() => {
    if (start) {
      window.addEventListener("mouseup", dragEnd);
      window.addEventListener("mousemove", dragging);
    }
  }, [start]);

  const containerStyles = styles && styles.container;
  const dragStyles = (styles && styles.drag) || {
    right: {},
    bottom: {},
    corner: {}
  };
  const DragSingleDirection =
    dragComponents && dragComponents.DragSingleDirection;
  const DragCorner = dragComponents && dragComponents.DragCorner;

  return (
    <Container
      id="test-container-resizer"
      style={{ width, height, ...containerStyles }}
    >
      <Drag
        variant="right"
        onMouseDown={e => onStartDrag(e, "vertical")}
        style={dragStyles.right}
      >
        {DragSingleDirection}
      </Drag>
      <Drag
        variant="bottom"
        onMouseDown={e => onStartDrag(e, "horizontal")}
        style={dragStyles.bottom}
      >
        {DragSingleDirection}
      </Drag>
      <Drag
        id="test-corner-drag"
        variant="corner"
        onMouseDown={e => onStartDrag(e, "both")}
        style={dragStyles.corner}
      >
        {DragCorner}
      </Drag>
      {children}
    </Container>
  );
};

export default Resizer;
