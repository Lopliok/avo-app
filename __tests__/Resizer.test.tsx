import * as React from "react";
import Enzyme, { shallow, ShallowWrapper } from "enzyme";
import { create, ReactTestRenderer } from "react-test-renderer";

import Resizer, { ResizerProps } from "../src/components/Resizer";

let wrapper: ShallowWrapper<ResizerProps>;
let snapshot: ReactTestRenderer;

beforeEach(() => {
  const resizer = (
    <Resizer
      defHeight={100}
      defWidth={300}
      styles={{
        container: { border: "1px solid black" },
        drag: { right: {}, bottom: {}, corner: {} }
      }}
    >
      <div style={{ background: "lightblue", height: "100%" }}>Example 1</div>
    </Resizer>
  );

  wrapper = shallow(resizer);
  snapshot = create(resizer);
});

describe("<Resizer />", () => {
  test("it matches the snapshot", () => {
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it("it should toggle checkbox label after click event", () => {
    expect(wrapper.text()).toEqual("Example 1");

    wrapper
      .find("#test-container-resizer")
      .simulate("mouseDown", { clientX: 0, clientY: 0 })
      .simulate("mouseMove", { clientX: 120, clientY: 100 })
      .simulate("mouseUp");
  });
});
