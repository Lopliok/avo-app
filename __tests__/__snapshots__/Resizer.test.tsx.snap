// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Resizer /> it matches the snapshot 1`] = `
<div
  className="sc-bdVaJa jGgSYQ"
  id="test-container-resizer"
  style={
    Object {
      "border": "1px solid black",
      "height": 100,
      "width": 300,
    }
  }
>
  <div
    className="sc-bwzfXH hnfraZ"
    onMouseDown={[Function]}
    style={Object {}}
  />
  <div
    className="sc-bwzfXH FpJEa"
    onMouseDown={[Function]}
    style={Object {}}
  />
  <div
    className="sc-bwzfXH kCdSJV"
    id="test-corner-drag"
    onMouseDown={[Function]}
    style={Object {}}
  />
  <div
    style={
      Object {
        "background": "lightblue",
        "height": "100%",
      }
    }
  >
    Example 1
  </div>
</div>
`;
